statmoments/benchmark/benchlib.py

```diff
@@ -49,34 +49,35 @@ def create_engine(self, tr_len, cl_len):
 def benchmark(benchset):
   number, repeat = 1, 1
 
-# Name        Implementation                     MB  tr_cnt   tr_len  cl_cnt   tr/sec   time res_time
-# bivar_m1    bivar_vtk(m(1, 1))                 50    5000      500       1      725    6.9      0.0
-# bivar_m1    bivar_vtk(m(1, 1))                124    5000     1000       1      132   37.7      0.0
-  print('{:12}{:30}{:>7}{:>8}{:>9}{:>8}{:>8}{:>7}{:>9}'.format(
-        'Kernel', 'Implementation', 'MB', 'tr_cnt', 'tr_len', 'cl_cnt', 'tr/sec', 'time', 'res_time'))
+  print('{:12}{:30}{:>7}{:>8}{:>9}{:>8}{:>10}{:>11}{:>10}{:>7}'.format(
+    'Kernel', 'Implementation', 'MB', 'tr_cnt', 'tr_len', 'cl_cnt', 'tr/sec', 'avg_upd_ms', 'avg_tt_ms', 'time_s'))
 
   for name, params in benchset:
-    for fctr, tr_count, tr_len, cl_count in product(*params):
-      tproc, ttest = [], []
+    for engine_factory, tr_count, tr_len, cl_count in product(*params):
+      update_times, ttest_times = [], []
+      assert repeat > 0
       for _ in range(repeat):
-        # Garbage collect memory after the prev iteration to free the existing memory
-        traces, engine = None, None
-        traces, classifiers = fctr.create_data(tr_count, tr_len, cl_count)
-        engine = fctr.create_engine(tr_len, cl_count)
+        traces, classifiers = engine_factory.create_data(tr_count, tr_len, cl_count)
+        engine = engine_factory.create_engine(tr_len, cl_count)
         for _ in range(number):
-          start = timer()
+          start = perf_counter()
           engine.update(traces, classifiers)  # Streaming: layout and accumulator
-          tproc.append(timer() - start)
+          update_times.append(perf_counter() - start)
 
-          start = timer()
-          for _ in ttests(engine):           # On-demand: t-tests
-            _ = None
-          ttest.append(timer() - start)
+          start = perf_counter()
+          for _ in ttests(engine):  # On-demand: t-tests
+            pass
+          ttest_times.append(perf_counter() - start)
 
-      mproc, mtt = min(tproc), min(ttest)
-      max_mom = ''.join(map(str, [engine.moment]*2))
+      total_time = sum(update_times) + sum(ttest_times)
+      max_mom = str(engine.moment) * 2
+      traces_count = tr_count * number
       kname = '{}(m{})'.format(type(engine._impl).__name__, max_mom)
-      print("{:12}{:30}{:7d}{:8d}{:9d}{:8d}{:>8d}{:>7.1f}{:>9.1f}".format(
-          name, kname, engine.memory_size >> 20, tr_count, tr_len, cl_count,
-          int(tr_count / mproc), mproc, mtt))
+      print("{:12}{:30}{:7d}{:8d}{:9d}{:8d}{:>10.1f}{:>11.3f}{:>10.3f}{:>7.1f}".format(
+        name, kname, engine.memory_size >> 20, traces_count, tr_len, cl_count, traces_count / total_time,
+                     sum(update_times) * 1000 / traces_count, sum(ttest_times) * 1000 / traces_count, total_time))
+      # Force garbage collection
+      del traces
+      del engine
+      gc.collect()
```

Comment:
Does `gc` really needed? `traces` and `engine` must be cleared by the time

Reply:
Yes, likely it is garbage collected. Just to make sure there is no other unnecessary stuff remaining.